modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/utils/getFields.ts

```diff
@@ -23,127 +28,110 @@ interface IProperty {
-						schemaName: propertyValue.items.$ref.replace(
-							/^.*\//,
-							''
-						),
-						schemas,
-					});
-				}
+			const field: IField = {
+				format: propertyValue.format,
+				label: propertyKey,
+				name: `${contextPath}${propertyKey}`,
+				type,
+			};
 
-				fields.push(field);
+			let targetSchemaName;
 
-				return;
+			if (propertyValue.items?.$ref) {
+				field.name = `${field.name}${FDS_ARRAY_FIELD_NAME_PARENT_SUFFIX}`;
+				field.type = type ? type : 'array';
+				targetSchemaName = propertyValue.items.$ref.replace(
+					/^.*\//,
+					''
+				);
 			}
-
-			if (propertyValue.$ref) {
-				fields.push({
-					children: getValidFields({
-						contextPath: `${contextPath}${propertyKey}${FDS_NESTED_FIELD_NAME_DELIMITER}`,
-						schemaName: propertyValue.$ref.replace(/^.*\//, ''),
-						schemas,
-					}),
-					label: propertyKey,
-					name: `${contextPath}${propertyKey}${FDS_NESTED_FIELD_NAME_PARENT_SUFFIX}`,
-					sortable: false,
-					type: type ? type : 'object',
-				});
-
-				return;
+			else if (propertyValue.$ref) {
+				field.name = `${field.name}${FDS_NESTED_FIELD_NAME_PARENT_SUFFIX}`;
+				field.type = type ? type : 'object';
+				targetSchemaName = propertyValue.$ref.replace(/^.*\//, '');
 			}
-
-			if (
+			else if (
 				propertyValue.extensions &&
 				propertyValue.extensions['x-parent-map'] === 'properties'
 			) {
 				const schemaNames = Object.keys(schemas);
-				const parentSchemaName = schemaNames.filter((schemaName) => {
+				const parentSchemaName = schemaNames.find((schemaName) => {
 					return (
 						schemaName.toLowerCase() ===
 						propertyKey.toLocaleLowerCase()
 					);
 				});
 
-				if (parentSchemaName.length) {
-					fields.push({
-						children: getValidFields({
-							contextPath: `${contextPath}${propertyKey}${FDS_NESTED_FIELD_NAME_DELIMITER}`,
-							schemaName: parentSchemaName[0],
-							schemas,
-						}),
-						label: propertyKey,
-						name: `${contextPath}${propertyKey}${FDS_NESTED_FIELD_NAME_PARENT_SUFFIX}`,
-						sortable: false,
-						type: schemas[parentSchemaName[0]]?.type || 'object',
-					});
-
-					return;
+				if (parentSchemaName) {
+					field.name = `${field.name}${FDS_NESTED_FIELD_NAME_PARENT_SUFFIX}`;
+					field.type = schemas[parentSchemaName]?.type || 'object';
+					targetSchemaName = parentSchemaName;
 				}
 			}
 
-			fields.push({
-				format: propertyValue.format,
-				label: propertyKey,
-				name: `${contextPath}${propertyKey}`,
-				sortable:
-					type !== 'object' &&
-					type !== 'array' &&
-					!contextPath.includes(FDS_NESTED_FIELD_NAME_DELIMITER) &&
-					!contextPath.includes(FDS_ARRAY_FIELD_NAME_DELIMITER),
-				type,
-			});
+			field.sortable =
+				type !== 'object' &&
+				type !== 'array' &&
+				!contextPath.includes(FDS_NESTED_FIELD_NAME_DELIMITER) &&
+				!contextPath.includes(FDS_ARRAY_FIELD_NAME_DELIMITER);
+
+			if (targetSchemaName && !schemaStack.includes(targetSchemaName)) {
```

Comment:
I would't call this `*stack`, as stack is a well known data structure. It could be misleading and confusing that it's not implemented in push/pop way. Maybe `visitedFields`, to match [visit()](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/FieldSelectModalContent.tsx#L22-L30)?